pkg/reconciler/rolloutorchestrator/rolloutorchestrator.go

```diff
@@ -0,0 +1,395 @@
+	if apierrs.IsNotFound(err) {
+		return r.createStagePA(ctx, ro, targetRev, scaleUpReady, UpdateSPAForRevDown)
+	}
+	if err != nil {
+		return spa, err
+	}
+	return r.client.ServingV1().StagePodAutoscalers(ro.Namespace).Update(ctx,
+		UpdateSPAForRevDown(spa, targetRev, scaleUpReady), metav1.UpdateOptions{})
+}
+
+// createOrUpdateSPARevUp create or update the StagePodAutoscaler, based on the specific (Stage)TargetRevision
+// defined in the RolloutOrchestrator for the revision scaling up.
+func (r *Reconciler) createOrUpdateSPARevUp(ctx context.Context, ro *v1.RolloutOrchestrator,
+	targetRev *v1.TargetRevision) (*v1.StagePodAutoscaler, error) {
+	spa, err := r.stagePodAutoscalerLister.StagePodAutoscalers(ro.Namespace).Get(targetRev.RevisionName)
+	if apierrs.IsNotFound(err) {
+		return r.createStagePA(ctx, ro, targetRev, true, UpdateSPAForRevUp)
+	}
+	if err != nil {
+		return spa, err
+	}
+	return r.client.ServingV1().StagePodAutoscalers(ro.Namespace).Update(ctx,
+		UpdateSPAForRevUp(spa, targetRev, true), metav1.UpdateOptions{})
+}
+
+// ReconcileKind implements Interface.ReconcileKind.
+func (r *Reconciler) ReconcileKind(ctx context.Context, ro *v1.RolloutOrchestrator) pkgreconciler.Event {
+	ctx, cancel := context.WithTimeout(ctx, pkgreconciler.DefaultTimeout)
+	defer cancel()
+
+	// If spec.StageRevisionStatus is nil, do nothing.
+	if len(ro.Spec.StageTargetRevisions) == 0 {
+		return nil
+	}
+
+	// Spec.StageTargetRevisions in the RolloutOrchestrator defines what the current stage looks like, in terms
+	// of the available revisions, and their name, traffic percentage, target number of replicas, whether it
+	// scales up or down, min and max scales defined by the Knative Service.
+	stageTargetRevisions := ro.Spec.StageTargetRevisions
+	revScalingUp, revScalingDown, err := RetrieveRevsUpDown(stageTargetRevisions)
+	if err != nil {
+		return err
+	}
+
+	// Create or update the StagePodAutoscaler for the revision scale up
+	for i := range revScalingUp {
+		if _, err = r.createOrUpdateSPARevUp(ctx, ro, revScalingUp[i]); err != nil {
+			return err
+		}
+	}
+
+	// If spec.StageRevisionStatus is nil, check on if the number of replicas meets the conditions.
+	if ro.IsStageInProgress() {
+		for i := range revScalingUp {
+			spa, err := r.stagePodAutoscalerLister.StagePodAutoscalers(ro.Namespace).Get(revScalingUp[i].RevisionName)
+			if err != nil {
+				return err
+			}
+
+			// spa.IsStageScaleInReady() returns true, as long as both DesireScale and ActualScale are available.
+			if !spa.IsStageScaleInReady() || !IsStageScaleUpReady(spa, revScalingUp[i]) {
+				// Create the stage pod autoscaler with the new maxScale set to
+				// maxScale defined in the revision traffic, because scale up phase is not over, we cannot
+				// scale down the old revision.
+				// Create or update the stagePodAutoscaler for the revision to be scaled down, eve if the scaling up
+				// phase is not over.
+				if len(revScalingDown) == 0 {
+					return nil
+				}
```

Comment:
nit: this code black is redundant

Reply:
No. It is possible that there is no revision running at this moment for the knative service, and we launch one revision for the service with 100% of the traffic assigned for the first time.

In this case, revScalingDown would be empty.